# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/webgpu/CommandEncoder.cpp
# Commit: a5be865d06b1
# Full Hash: a5be865d06b167e20cff16ad33e4f076bbe41757
# Author: Erich Gubler <erichdongubler@gmail.com>
# Date: 2025-07-17 09:30:54
# Regressor Bug: 1976960
# File Overlap Count: 1
# Description:
#   Bug 1976960 - feat(webgpu): allow `GPUTexture`s where `GPUTextureView`s are allowed r=webgpu-reviewers,webidl,smaug,nical
#   
#   Changed in spec. upstream by [gpuweb#5228](https://github.com/gpuweb/gpuweb/pull/5228), tested by CTS in [gpuweb/cts#4407](https://github.com/gpuweb/cts/pull/4407) (which is what caused the `FAIL`s removed here to be introduced with bug 1976874).
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D257054
# ==============================================================================

diff -r e98afaa9f97b -r a5be865d06b1 dom/webgpu/CommandEncoder.cpp
--- a/dom/webgpu/CommandEncoder.cpp	Thu Jul 17 04:08:37 2025 +0000
+++ b/dom/webgpu/CommandEncoder.cpp	Thu Jul 17 04:08:38 2025 +0000
@@ -12,6 +12,7 @@
 #include "ComputePassEncoder.h"
 #include "Device.h"
 #include "RenderPassEncoder.h"
+#include "TextureView.h"
 #include "Utility.h"
 #include "mozilla/webgpu/CanvasContext.h"
 #include "mozilla/webgpu/ffi/wgpu.h"
@@ -222,9 +223,29 @@
   dom::GPURenderPassDescriptor desc{aDesc};
 
   for (auto& at : desc.mColorAttachments) {
-    TrackPresentationContext(at.mView->GetTargetContext());
+    auto coerceToViewInPlace =
+        [](dom::OwningGPUTextureOrGPUTextureView& texOrView)
+        -> RefPtr<TextureView> {
+      RefPtr<TextureView> view;
+      switch (texOrView.GetType()) {
+        case dom::OwningGPUTextureOrGPUTextureView::Type::eGPUTexture: {
+          dom::GPUTextureViewDescriptor defaultDesc{};
+          RefPtr<Texture> tex = texOrView.GetAsGPUTexture();
+          texOrView.SetAsGPUTextureView() = tex->CreateView(defaultDesc);
+          break;
+        }
+
+        case dom::OwningGPUTextureOrGPUTextureView::Type::eGPUTextureView:
+          // Nothing to do, great!
+          break;
+      }
+      view = texOrView.GetAsGPUTextureView();
+      return view;
+    };
+    TrackPresentationContext(coerceToViewInPlace(at.mView)->GetTargetContext());
     if (at.mResolveTarget.WasPassed()) {
-      TrackPresentationContext(at.mResolveTarget.Value().GetTargetContext());
+      TrackPresentationContext(
+          coerceToViewInPlace(at.mResolveTarget.Value())->GetTargetContext());
     }
   }
 